### PR TITLE
Conditional vs marginal importance on the XOR dataset

### PR DESCRIPTION
**Pull request by jpaillard**
- This issue is related to #160 . This PR does not address all the points mentioned in the discussion but illustrates a special case where the marginal importance is null but the conditional importance is not. 
 - It is inspired by the sklearn example on the same data: https://scikit-learn.org/stable/auto_examples/gaussian_process/plot_gpc_xor.html
 - The example uses the XOR dataset and SVC classifier. The conditional importance is estimated using CPI and the marginal using univariate models.

### CODE REVIEW

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+plt.show()
+
+
+###############################################################################
+# Computing the conditional and marginal importance
+# -----------------------------------------------------
+# We first compute the marginal importance by fitting univariate models on each feature.
+# Then, we compute the conditional importance using the CPI class. The univarariate
+# models don't perform above chance, since solving the XOR problem requires to use both
+# features. The conditional importance, on the other hand, reveals that both features
```
Comment (bthirion) [inline examples/plot_conditional_vs_marginal_xor_data.py:70]:
```suggestion
# features. Conditional importance, on the other hand, reveals that both features
```

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        estimator=clf_c,
+        method="decision_function",
+        loss=hinge_loss,
+        imputation_model_continuous=RidgeCV(np.logspace(-3, 3, 10)),
+        n_permutations=50,
+        random_state=0,
+    )
+    vim.fit(X_train_cv, y_train_cv)
+    importances.append(vim.score(X_test_cv, y_test_cv)["importance"])
+importances = np.array(importances).T
```
Comment (bthirion) [inline examples/plot_conditional_vs_marginal_xor_data.py:111]: skip a line here

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    orient="h",
+    ax=axes[0],
+    fill=False,
+    color="C0",
+    linewidth=3,
+)
+axes[0].axvline(x=0.5, color="k", linestyle="--", lw=3)
+axes[0].set_ylabel("")
+axes[0].set_yticks([0, 1], ["X1", "X2"])
+axes[0].set_xlabel("Marginal Scores (accuracy)")
```
Comment (bthirion) [inline examples/plot_conditional_vs_marginal_xor_data.py:126]: I can't see the ax labels on the generated figure

Comment (jpaillard) [reply examples/plot_conditional_vs_marginal_xor_data.py:128]: `plt.tight_layout` seems to have solved the issue.

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+)
+
+X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=0)
+model = SVC(kernel="rbf", random_state=0)
+model.fit(X_train, y_train)
+
+Z = model.decision_function(np.c_[xx.ravel(), yy.ravel()])
+fig, ax = plt.subplots()
+
+ax.imshow(
```
Comment (lionelkusch) [inline examples/plot_conditional_vs_marginal_xor_data.py:40]: If you want to do this plot for explain the data. In my opinion, it's better to use a density plot (see the function: https://seaborn.pydata.org/generated/seaborn.kdeplot.html).

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
+"""
+Conditional vs Marginal Importance on the XOR dataset
+==================================================================
+
+This example illustrates on XOR data that variables can be conditionally important even
+if they are not marginally important. The conditional importance is computed using the
+CPI class and the marginal importance is computed using univariate models.
+"""
+
+#############################################################################
```
Comment (lionelkusch) [inline examples/plot_conditional_vs_marginal_xor_data.py:10]: Can you put this line after the import?

In my view, it's better to have a cell with just the import.

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    np.linspace(np.min(X[:, 0]), np.max(X[:, 0]), 100),
+    np.linspace(np.min(X[:, 1]), np.max(X[:, 1]), 100),
+)
+
+X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=0)
+model = SVC(kernel="rbf", random_state=0)
+model.fit(X_train, y_train)
+
+Z = model.decision_function(np.c_[xx.ravel(), yy.ravel()])
+fig, ax = plt.subplots()
```
Comment (lionelkusch) [inline examples/plot_conditional_vs_marginal_xor_data.py:38]: In my opinion, it's better to create a cell just for the plot of data.

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+)
+ax.axis("off")
+ax.legend(
+    title="Class",
+)
+ax.set_title("Decision function of SVC with RBF kernel")
+plt.show()
+
+
+###############################################################################
```
Comment (lionelkusch) [inline examples/plot_conditional_vs_marginal_xor_data.py:64]: In my opinion, it's better to create a cell by methods or to create a function.

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        imputation_model_continuous=RidgeCV(np.logspace(-3, 3, 10)),
+        n_permutations=50,
+        random_state=0,
+    )
+    vim.fit(X_train, y_train)
+    importances.append(vim.score(X_test, y_test)["importance"])
+
+importances = np.array(importances).T
+
+# Plotting the results
```
Comment (lionelkusch) [inline examples/plot_conditional_vs_marginal_xor_data.py:114]: You should create a specific cell for the plot.

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    orient="h",
+    ax=axes[1],
+    fill=False,
+    color="C0",
+    linewidth=3,
+)
+axes[1].set_xlabel("Conditional Importance")
+axes[1].axvline(x=0.0, color="k", linestyle="--", lw=3)
+plt.tight_layout()
+plt.show()
```
Comment (lionelkusch) [inline examples/plot_conditional_vs_marginal_xor_data.py:143]: Can you add an interpretation of the figure?

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    hue=Y,
+    ax=ax,
+    palette="muted",
+)
+ax.axis("off")
+ax.legend(
+    title="Class",
+)
+ax.set_title("Decision function of SVC with RBF kernel")
+plt.show()
```
Comment (lionelkusch) [inline examples/plot_conditional_vs_marginal_xor_data.py:61]: Can you add an interpretation of the figure?

File: examples/plot_conditional_vs_marginal_xor_data.py (hunk 4 above)
Comment (jpaillard) [reply examples/plot_conditional_vs_marginal_xor_data.py:40]: I thought about it initially but agreed that level lines don't add much information and incur additional computation costs. The point is to illustrate that the SCV solves the classification task visually I thought imshow is good enough for that. WDYT?

File: examples/plot_conditional_vs_marginal_xor_data.py (hunk 10 above)
Comment (jpaillard) [reply examples/plot_conditional_vs_marginal_xor_data.py:62]: done

File: examples/plot_conditional_vs_marginal_xor_data.py (hunk 9 above)
Comment (jpaillard) [reply examples/plot_conditional_vs_marginal_xor_data.py:144]: done

File: examples/plot_conditional_vs_marginal_xor_data.py (hunk 4 above)
Comment (lionelkusch) [reply examples/plot_conditional_vs_marginal_xor_data.py:41]: I prefer to have the explanation attached to the figure that before the code.
If you can add a name for this section, it will be nice.

Comment (jpaillard) [reply examples/plot_conditional_vs_marginal_xor_data.py:43]: done

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        loss=hinge_loss,
+        imputation_model_continuous=RidgeCV(np.logspace(-3, 3, 10)),
+        n_permutations=50,
+        random_state=0,
+    )
+    vim.fit(X_train, y_train)
+    importances.append(vim.importance(X_test, y_test)["importance"])
+
+importances = np.array(importances).T
+
```
Comment (lionelkusch) [inline examples/plot_conditional_vs_marginal_xor_data.py:123]: You should create a separation here and add a"numerical" presentation of the result.

Comment (jpaillard) [reply examples/plot_conditional_vs_marginal_xor_data.py:123]: done

**Review by bthirion — APPROVED** [review]
LGTM, thx !

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
+"""
+Conditional vs Marginal Importance on the XOR dataset
+==================================================================
+
+This example illustrates on XOR data that variables can be conditionally important even
+if they are not marginally important. The conditional importance is computed using the
+CPI class and the marginal importance is computed using univariate models.
```
[the remark this answers is not in the document]
Comment (lionelkusch) [reply examples/plot_conditional_vs_marginal_xor_data.py:7]: If you use some acronym, write the entire words before using it.

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    np.linspace(np.min(X[:, 1]), np.max(X[:, 1]), 100),
+)
+
+X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=0)
+model = SVC(kernel="rbf", random_state=0)
+model.fit(X_train, y_train)
+Z = model.decision_function(np.c_[xx.ravel(), yy.ravel()])
+
+#############################################################################
+#  Visualizing the decision function of the SVC
```
[the remark this answers is not in the document]
Comment (lionelkusch) [reply examples/plot_conditional_vs_marginal_xor_data.py:39]: ?

File: examples/plot_conditional_vs_marginal_xor_data.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+##############################################################################
+# The decision function of the SVC shows that the model is able to learn the
+# non-linear decision boundary of the XOR problem. It also highlights that knowing
+# the value of both features is necessary to classify each sample correctly.
+
+
+###############################################################################
+# Computing the conditional and marginal importance
+# -----------------------------------------------------
```
Comment (lionelkusch) [inline examples/plot_conditional_vs_marginal_xor_data.py:75]:
```suggestion
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Review by lionelkusch — APPROVED** [review]
